engine/src/cmd/damageable.cpp

```diff
@@ -45,242 +53,484 @@ static float  flickeronprob = 0.66f; // ( "graphics", "glowflicker", "num-times-
+    bool is_restricted = restricted_items.find(cargo.GetContent()) == string::npos;
+
+    // The following comment was kept in the hopes someone else knows what it means
+    //why not downgrade _add GetCargo(which).content.find("add_")!=0&&
+    if(!is_upgrade || already_damaged || is_multiple || is_restricted) {
+        return;
+    }
+
+
+    const int prefix_length = strlen( "upgrades/" );
+    cargo.category = "upgrades/Damaged/" + cargo_category.substr(prefix_length);
+
+    // TODO: find a better name for whatever this is. Right now it's not not downgrade
+    static bool not_actually_downgrade = GameConfig::GetVariable( "physics", "separate_system_flakiness_component", false );
+    if (not_actually_downgrade) {
+        return;
+    }
+
+    const Unit *downgrade = loadUnitByCache( cargo.content, FactionUtil::GetFactionIndex( "upgrades" ) );
+    if (!downgrade) {
+        return;
+    }
+
+    if ( 0 == downgrade->getNumMounts() && downgrade->SubUnits.empty() ) {
+        double percentage = 0;
+        unit->Downgrade( downgrade, 0, 0, percentage, nullptr );
+    }
+}
+
+
+// TODO: what else is needed here that we have a separate function?
+void Damageable::Destroy() {
+    DamageableObject::Destroy();
+}
+
+
+float Damageable::FShieldData() const
+{
+    Damageable *damageable = const_cast<Damageable*>(this);
+    return static_cast<DamageableLayer>(damageable->GetShieldLayer()).GetPercent(FacetName::front);
+}
+
+float Damageable::BShieldData() const
+{
+    Damageable *damageable = const_cast<Damageable*>(this);
+    return static_cast<DamageableLayer>(damageable->GetShieldLayer()).GetPercent(FacetName::rear);
+}
+
+float Damageable::LShieldData() const
+{
+    Damageable *damageable = const_cast<Damageable*>(this);
+    return static_cast<DamageableLayer>(damageable->GetShieldLayer()).GetPercent(FacetName::left);
+}
+
+float Damageable::RShieldData() const
+{
+    Damageable *damageable = const_cast<Damageable*>(this);
+    return static_cast<DamageableLayer>(damageable->GetShieldLayer()).GetPercent(FacetName::right);
 }
 
 //short fix
 void Damageable::ArmorData( float armor[8] ) const
 {
-    armor[0] = this->armor.frontrighttop;
-    armor[1] = this->armor.backrighttop;
-    armor[2] = this->armor.frontlefttop;
-    armor[3] = this->armor.backlefttop;
-    armor[4] = this->armor.frontrightbottom;
-    armor[5] = this->armor.backrightbottom;
-    armor[6] = this->armor.frontleftbottom;
-    armor[7] = this->armor.backleftbottom;
+    Damageable *damageable = const_cast<Damageable*>(this);
+    DamageableLayer armor_layer = damageable->GetArmorLayer();
+    armor[0] = armor_layer.facets[0].health;
+    armor[1] = armor_layer.facets[1].health;
+    armor[2] = armor_layer.facets[2].health;
+    armor[3] = armor_layer.facets[3].health;
+    armor[4] = armor_layer.facets[4].health;
+    armor[5] = armor_layer.facets[5].health;
+    armor[6] = armor_layer.facets[6].health;
+    armor[7] = armor_layer.facets[7].health;
```

Comment:
might a for-loop do better here:
```
for int i = 0; i < 8; i++ {
    armor[i] = armor_layer.facets[i].health;
}
```
the compiler might unroll it to what you have, but it's just as effective and clear.

Reply:
You are absolutely right. I suspect I originally needed to implement 

`armor[i] = armor_layer.facets[j].health;

where i != j and fixed it along the way. Honestly, I no longer remember.